tokio/src/net/unix/pipe.rs

```diff
@@ -0,0 +1,1207 @@
+    ///
+    /// Fails with `io::ErrorKind::InvalidInput` if the file is not a pipe or it
+    /// does not have read access. Also fails with any standard OS error if it occurs.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn from_file(mut file: File) -> io::Result<Receiver> {
+        if !is_fifo(&file)? {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "not a pipe"));
+        }
+
+        let flags = get_file_flags(&file)?;
+        if has_read_access(flags) {
+            set_nonblocking(&mut file, flags)?;
+            Receiver::from_file_unchecked(file)
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "not in O_RDONLY or O_RDWR access mode",
+            ))
+        }
+    }
+
+    /// Creates a new `Receiver` from a [`File`] without checking pipe properties.
+    ///
+    /// This function is intended to construct a pipe from a File representing
+    /// a special FIFO file. The conversion assumes nothing about the underlying
+    /// file; it is left up to the user to make sure it is opened with read access,
+    /// represents a pipe and is set in non-blocking mode.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::net::unix::pipe;
+    /// use std::fs::OpenOptions;
+    /// use std::os::unix::fs::{FileTypeExt, OpenOptionsExt};
+    /// # use std::error::Error;
+    ///
+    /// const FIFO_NAME: &str = "path/to/a/fifo";
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn Error>> {
+    /// let file = OpenOptions::new()
+    ///     .read(true)
+    ///     .custom_flags(libc::O_NONBLOCK)
+    ///     .open(FIFO_NAME)?;
+    /// if file.metadata()?.file_type().is_fifo() {
+    ///     let rx = pipe::Receiver::from_file_unchecked(file)?;
+    ///     /* use the Receiver */
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn from_file_unchecked(file: File) -> io::Result<Receiver> {
+        let raw_fd = file.into_raw_fd();
+        let mio_rx = unsafe { mio_pipe::Receiver::from_raw_fd(raw_fd) };
+        Receiver::from_mio(mio_rx)
+    }
+
+    /// Waits for any of the requested ready states.
+    ///
+    /// This function can be used to wait for [`READABLE`] and [`READ_CLOSED`] events.
```

Comment:
The `Interest` type does not define a `READ_CLOSED` constant. Similarly for the sender.

Reply:
I meant waiting for [`Ready::READ_CLOSED`](https://docs.rs/tokio/1.25.0/tokio/io/struct.Ready.html#associatedconstant.READ_CLOSED) with `Interest::READABLE`. I added this method in addition to `readable` only to allow detecting when a pipe is closed and not readable, because `readable` does not return the actual `Ready` set. I can add a note about it here.